src/genericitems/producers.h

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2021 Team Dissolve and contributors
+
+#pragma once
+
+#include <any>
+#include <functional>
+#include <string>
+#include <typeindex>
+#include <unordered_map>
+
+// GenericItem Producer
+class GenericItemProducer
+{
+    private:
+    GenericItemProducer();
+
+    public:
+    GenericItemProducer(const GenericItemProducer &) = delete;
+    GenericItemProducer &operator=(const GenericItemProducer &) = delete;
+
+    /*
+     * Producers
+     */
+    private:
+    // Producer function type
+    using ProducerFunction = std::function<std::any()>;
+    // Producers for all data types
+    std::unordered_map<std::type_index, std::string> classNames_;
+    std::unordered_map<std::type_index, ProducerFunction> producers_;
+
+    private:
+    // Register producer for specific class
+    template <class T> void registerProducer(std::string className)
+    {
+        producers_[typeid(T)] = []() { return T(); };
+        classNames_[typeid(T)] = std::move(className);
+    }
```

Comment:
```suggestion
    template <class T> void registerProducer()
    {
        producers_[typeid(T)] = []() { return T(); };
        classNames_[typeid(T)] = std::move(typeid(T).name());
    }
```

This would remove the need to give the type name in both the template and the string for every type.  It does remove the ability to register a producer under the "wrong" name, but I didn't find an instance of using that.

Reply:
Unfortunately, since type names derived from `type_info` are implementation-dependent and often mangled, this would break restart-file compatibility across platforms. So, for consistency the manual naming approach has to remain.

Reply:
Good call.  I haven't considered the cross platform issues.